src/repo/remote.rs

```diff
@@ -0,0 +1,692 @@
+use std::{
+    convert::TryInto,
+    ffi::OsStr,
+    fs, io,
+    io::{BufRead, BufReader, Read, Write},
+    path::{Path, PathBuf},
+    time::{Duration, SystemTime},
+};
+
+use chrono::{offset::Utc, DateTime};
+use crypto::digest::Digest;
+use crypto::sha1::Sha1;
+use ureq::AgentBuilder;
+use url::Url;
+
+use super::constants::{PACK_EXTENSION, REMOTE_INDEX_EXTENSION, REMOTE_INDEX_SIZE_LIMIT};
+use super::error::Error;
+use super::fs::{create_file, open_file};
+use crate::packidx::{ObjectChecksum, PackIndex};
+use crate::progress::{ProgressReporter, ProgressWriter};
+
+const HTTP_STATUS_OK: u16 = 200;
+const HTTP_STATUS_NOT_MODIFIED: u16 = 304;
+
+trait ReifyRemoteResult {
+    fn reify<D: std::fmt::Display>(self, display_name: D) -> Self;
+}
+
+impl<T> ReifyRemoteResult for Result<T, Error> {
+    fn reify<D: std::fmt::Display>(self, display_name: D) -> Result<T, Error> {
```

Comment:
This is a pretty interesting pattern, to have something produce a value with an empty slot which you then annotate with detail later. I haven't seen it so far, so it's new to me. I haven't decided if it makes sense to me yet.

The thing that it competes with is something which wraps the error value with the extra information instead. I feel this is the more obvious thing to do, and that having a 'slot' which you fill is a weird way to handle the state as opposed to just having 'the error value', and then putting that error value into such a slot when the slot is needed. Something like this instead: `Error::RemoteError(remote_path, err)`.

Reply:
Additional thought: you might still be able to keep reify trait as a convenience, but just have the implementation wrap the error rather than fill a slot as it currently does.